KAFKA-3616: Make kafka producers/consumers injectable for KafkaStreams

kawamuray opened the pull request:
Ticket: https://issues.apache.org/jira/browse/KAFKA-3616

kawamuray committed:
KAFKA-3616: Make kafka producers/consumers injectable for KafkaStreams

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -98,16 +98,21 @@
+        this(builder, new StreamsConfig(props), new DefaultKafkaClientSupplier());
+    }
+
+    public KafkaStreams(TopologyBuilder builder, StreamsConfig config) {
```
Good catch!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -98,16 +98,21 @@
      *
      * @param builder The processor topology builder specifying the computational logic
      * @param config The stream configs
+     * @param clientSupplier The kafka clients supplier which provides clients for this KafkaStream
```
nit: "which provide underlying producer and consumer clients for this {@code KafkaStream} instance".

kawamuray replied:
ack. maybe `@link` though? Fixed by:    8f45081

guozhangwang commented:
@kawamuray Could you rebase?

A couple of general comments:

1) Move DefaultClientSupplier into streams/processor/internals as it does not need to be public interface.
2) Could you also create a MockClientSupplier which uses the MockProducer and MockConsumer and add a unit test to validate that the clients are generated as expected?

kawamuray committed:
(fixup) Add MockClientSupplier and fix unit tests to adjust constructor signature change

kawamuray committed:
(fixup) Add a unit test to verify that the clients are truly generated through KafkaClientSupplier

kawamuray committed:
MINOR: Cleanup unused variables on unit tests

kawamuray committed:
(fixup) format javadocs

kawamuray commented:
@guozhangwang Thanks for reviewing, rebased and addressed your comments with following up unit test fixes.

For your two comments:

> 1) Move DefaultClientSupplier into streams/processor/internals as it does not need to be public interface.

Actually I was thinking below as the practical use since I'm expecting that in most cases only producer or consumer is the injection target. What do you think of?

``` java
KafkaStreams streams = new KafkaStreams(topology, config, new DefaultKafkaClientSupplier() {
    @Override
    public Producer<byte[], byte[]> getProducer(Map<String, Object> config) {
        return new MyKafkaProducer<>(...);
    }
});
```

> 2) Could you also create a MockClientSupplier which uses the MockProducer and MockConsumer and add a unit test to validate that the clients are generated as expected?

Ack. Addressed by: c2a5a5a and 8b13ced.

kawamuray committed:
(fixup) fix javadoc as suggested by review

kawamuray committed:
(fixup) restore javadoc of traditional constructor

kawamuray committed:
(fixup) indent

guozhangwang commented:
LGTM overall, Jenkins failure are not related (filed separate JIRA).

About the DefaultClientSupplier: in general I'd prefer to keep things private until we see common requirement for them to be public, since vice versa cannot be easily done due to backward incompatibility. Thoughts?

kawamuray commented:
> About the DefaultClientSupplier: in general I'd prefer to keep things private until we see common requirement for them to be public, since vice versa cannot be easily done due to backward incompatibility. Thoughts?

Okay, as all methods of DefaultKafkaClientSupplier are fairly easy to re-implement by users, let me follow your advice ATM :+1:

kawamuray committed:
(fixup) Move defaultKafkaClientSupplier to under the internals

guozhangwang commented:
Merged to trunk and cherry-picked to 0.10.0.
